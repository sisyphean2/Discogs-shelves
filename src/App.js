import { useCallback, useEffect, useReducer, useState } from 'react';

import { Container, Grid } from '@material-ui/core';

import { DragDropContext } from 'react-beautiful-dnd';

import { reducer } from './reducer';

import RecordsContainer from './components/RecordsContainer';
import Shelves from './components/Shelves';

export default function App() {
  const [records, setRecords] = useState([]);

  const [shelves, dispatch] = useReducer(reducer, {});

  const onDragEnd = useCallback(
    result => {
      const { source, destination } = result;

      if (!destination) {
        return;
      }

      if (source.droppableId === destination.droppableId) {
        dispatch({
          type: 'reorderInShelf',
          shelfId: source.droppableId,
          oldIndex: source.index,
          newIndex: destination.index,
        });
      } else {
        dispatch({
          type: 'moveBetweenShelves',
          oldShelf: source.droppableId,
          newShelf: destination.droppableId,
          oldIndex: source.index,
          newIndex: destination.index,
        });
      }
    },
    [dispatch],
  );

  useEffect(() => {
    fetch(
      'https://api.discogs.com/users/blacklight/collection/folders/0/releases',
    )
      .then(resp => resp.json())
      .then(json => {
        setRecords(
          json.releases.map(release => {
            const { id, basic_information: info } = release;
            return {
              id: `${id}`,
              title: info.title,
              formats: info.formats.map(format => format.name),
              label: info.labels?.[0]?.name ?? '',
              artists: info.artists.map(artist => artist.name),
              date: info.year,
            };
          }),
        );
      });
  }, []);

  return (
    <Container>
      <h1>Record Shelves App</h1>

      <Grid container spacing={3}>
        <Grid item xs={3}>
          <RecordsContainer
            records={records}
            shelves={shelves}
            dispatch={dispatch}
          />
        </Grid>

        <Grid item xs={9}>
          <DragDropContext onDragEnd={onDragEnd}>
            <Shelves records={records} shelves={shelves} dispatch={dispatch} />
          </DragDropContext>
        </Grid>
      </Grid>
    </Container>
  );
}
