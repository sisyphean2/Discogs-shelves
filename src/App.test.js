import { fireEvent, render, screen, waitFor, within } from '@testing-library/react';

import { enableFetchMocks } from 'jest-fetch-mock';

import App from './App';

enableFetchMocks();

describe('App', () => {
  beforeEach(() => {
    fetch.resetMocks();
    fetch.doMock();
    fetch.mockResponseOnce(
      JSON.stringify({
        releases: [
          {
            id: 76035,
            basic_information: {
              master_id: 64128,
              title: 'Dirty Dancing',
              year: 2002,
              formats: [{ name: 'Vinyl' }],
              labels: [{ name: '!K7 Records' }],
              artists: [{ name: 'Swayzak' }],
            },
          },
        ],
      }),
    );
  });

  it('can add a shelf', async () => {
    const { asFragment } = render(<App />);
    await waitFor(() => expect(fetch).toHaveBeenCalledTimes(1));

    fireEvent.click(screen.getByText('Add Shelf'));
    fireEvent.change(screen.getByTestId('add-shelf'), {
      target: { value: 'first shelf' },
    });
    fireEvent.click(screen.getByText('Submit'));

    expect(asFragment()).toMatchSnapshot();
  });

  it('can remove a shelf', async () => {
    const { asFragment } = render(<App />);
    await waitFor(() => expect(fetch).toHaveBeenCalledTimes(1));

    fireEvent.click(screen.getByText('Add Shelf'));
    fireEvent.change(screen.getByTestId('add-shelf'), {
      target: { value: 'first shelf' },
    });
    fireEvent.click(screen.getByText('Submit'));

    fireEvent.click(screen.getByText('Remove'));
    expect(asFragment()).toMatchSnapshot();
  });
});
