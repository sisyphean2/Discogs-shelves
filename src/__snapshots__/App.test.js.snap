// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App can add a shelf 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg"
  >
    <h1>
      Record Shelves App
    </h1>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
      >
        <h2>
          Records
        </h2>
        <ul
          class="MuiList-root MuiList-padding"
          style="background-color: rgb(245, 245, 245); height: calc(100vh - 12rem); overflow: scroll;"
        >
          <li
            class="MuiListItem-root MuiListItem-gutters"
          >
            <div
              class="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
              style="width: 260px;"
            >
              <div
                class="MuiCardContent-root"
              >
                <p>
                  Title: Dirty Dancing
                </p>
                <p>
                  Artist(s):Swayzak
                </p>
                <p>
                  Label: !K7 Records
                </p>
                <p>
                  Formats: Vinyl
                </p>
              </div>
              <div
                class="MuiCardActions-root MuiCardActions-spacing"
              >
                <div
                  class="MuiFormControl-root"
                  style="min-width: 120px;"
                >
                  <label
                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
                    data-shrink="false"
                  >
                    Add to shelf
                  </label>
                  <div
                    class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
                    data-testid="add-shelf"
                  >
                    <div
                      aria-haspopup="listbox"
                      class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
                      role="button"
                      tabindex="0"
                    >
                      <span>
                        â€‹
                      </span>
                    </div>
                    <input
                      aria-hidden="true"
                      class="MuiSelect-nativeInput"
                      tabindex="-1"
                      value=""
                    />
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSelect-icon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M7 10l5 5 5-5z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </li>
        </ul>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-9"
      >
        <h2>
          Shelves
        </h2>
        <ul
          class="MuiList-root MuiList-padding"
          style="height: calc(100vh - 12rem); overflow: scroll;"
        >
          <div
            class="MuiBox-root MuiBox-root-1"
            style="display: flex; align-items: center; justify-content: space-between;"
          >
            <h3
              style="cursor: pointer;"
            >
              first shelf
            </h3>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined"
              style="margin-left: 1rem;"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                Remove
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
          <li
            class="MuiListItem-root MuiListItem-gutters"
            style="background-color: rgb(245, 245, 245); min-height: 10rem; margin-bottom: 1rem;"
          >
            <ul
              class="MuiList-root MuiList-padding"
              style="display: flex; flex-direction: row; overflow: scroll;"
            >
              <p>
                No records in shelf
              </p>
            </ul>
          </li>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Add Shelf
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </ul>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`App can remove a shelf 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg"
  >
    <h1>
      Record Shelves App
    </h1>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
      >
        <h2>
          Records
        </h2>
        <ul
          class="MuiList-root MuiList-padding"
          style="background-color: rgb(245, 245, 245); height: calc(100vh - 12rem); overflow: scroll;"
        >
          <li
            class="MuiListItem-root MuiListItem-gutters"
          >
            <div
              class="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
              style="width: 260px;"
            >
              <div
                class="MuiCardContent-root"
              >
                <p>
                  Title: Dirty Dancing
                </p>
                <p>
                  Artist(s):Swayzak
                </p>
                <p>
                  Label: !K7 Records
                </p>
                <p>
                  Formats: Vinyl
                </p>
              </div>
              <div
                class="MuiCardActions-root MuiCardActions-spacing"
              />
            </div>
          </li>
        </ul>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-9"
      >
        <h2>
          Shelves
        </h2>
        <ul
          class="MuiList-root MuiList-padding"
          style="height: calc(100vh - 12rem); overflow: scroll;"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Add Shelf
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </ul>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`App fetches records 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg"
  >
    <h1>
      Record Shelves App
    </h1>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
      >
        <h2>
          Records
        </h2>
        <ul
          class="MuiList-root MuiList-padding"
          style="background-color: rgb(245, 245, 245); height: calc(100vh - 12rem); overflow: scroll;"
        >
          <li
            class="MuiListItem-root MuiListItem-gutters"
          >
            <div
              class="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
              style="width: 260px;"
            >
              <div
                class="MuiCardContent-root"
              >
                <p>
                  Title: Dirty Dancing
                </p>
                <p>
                  Artist(s):Swayzak
                </p>
                <p>
                  Label: !K7 Records
                </p>
                <p>
                  Formats: Vinyl
                </p>
              </div>
              <div
                class="MuiCardActions-root MuiCardActions-spacing"
              />
            </div>
          </li>
        </ul>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-9"
      >
        <h2>
          Shelves
        </h2>
        <ul
          class="MuiList-root MuiList-padding"
          style="height: calc(100vh - 12rem); overflow: scroll;"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Add Shelf
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </ul>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
